Add Template for .NET 8 Wep App

mahald opened the pull request:
Same as the new BLazor Web App Template

mahald committed:
Add Template for .NET 8 Wep App

Same as the new BLazor Web App Template

mahald commented:
# New Web App Template for .NET 8

MudBlazor Web App template copied from Microsoft Web App template and modified to use MudBlazor components.
Requires .NET 8 latest SDK to be installed.
Requires latest Visual Studio 2022 Preview to be seen in the list of templates.
install the latest MudBlazor.Templates from NuGet to get the template.

## Tested Modes for new .NET 8 Web App Template

### Interactive per Page:

``` 
dotnet new MudBlazorWebApp -int (Auto|Server|WebAssembly)
```

### Interactive Global:

```
dotnet new MudBlazorWebApp -int (Auto|Server|WebAssembly) -ai
```

### Add Authentication:

```
dotnet new MudBlazorWebApp -int Auto -au Individual
```

## ToDo for the .NET 8 Template:
* Authentication not yet using MudBlazor components, need to be updated.
* Add more meaningful demo pages with actual fetching of data from server.
* Add more documentation on how to use the template.
* Currently Authentication is only working without the -ai option, need to fix that (should get fixed when resync with Blazor templates at .Net 8 launch in 3 weeks).

mahald committed:
Correcting README.md caption in doc for .net8 template

mahald commented:
To install for now: 

git clone https://github.com/mahald/Templates.git
dotnet new install ./Templates/src/MudBlazorWebTemplate

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "author": "Microsoft",
+  "name": "Aplicativo Web Blazor",
+  "description": "Um modelo de projeto para criar um aplicativo Web Blazor que dá suporte à renderização do lado do servidor e à interatividade do cliente. Este modelo pode ser usado para aplicativos da Web com interfaces de usuário (UIs) dinâmicas avançadas.",
+  "symbols/Framework/description": "A estrutura de destino do projeto.",
```
Hello @mahald , nice job here, congratulations. I'm going justo to leave some tips for better pt-BR translation:

"Estrutura de destino do projeto"

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "name": "Aplicativo Web Blazor",
+  "description": "Um modelo de projeto para criar um aplicativo Web Blazor que dá suporte à renderização do lado do servidor e à interatividade do cliente. Este modelo pode ser usado para aplicativos da Web com interfaces de usuário (UIs) dinâmicas avançadas.",
+  "symbols/Framework/description": "A estrutura de destino do projeto.",
+  "symbols/Framework/choices/net8.0/description": "net8.0 de destino",
```
.NET 8.0

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/Framework/description": "A estrutura de destino do projeto.",
+  "symbols/Framework/choices/net8.0/description": "net8.0 de destino",
+  "symbols/UserSecretsId/description": "O ID a ser usado para segredos (usado com autenticação individual).",
+  "symbols/skipRestore/description": "Se especificado, ignora a restauração automática do projeto sendo criado.",
```
Caso especificado, ignora a restauração automática do projeto sendo criado.

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/Framework/choices/net8.0/description": "net8.0 de destino",
+  "symbols/UserSecretsId/description": "O ID a ser usado para segredos (usado com autenticação individual).",
+  "symbols/skipRestore/description": "Se especificado, ignora a restauração automática do projeto sendo criado.",
+  "symbols/ExcludeLaunchSettings/description": "Se deve excluir launchSettings.json do modelo gerado.",
```
Define se launchSettings.json deve ser excluído do modelo gerado.

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/InteractivityPlatform/choices/WebAssembly/displayName": "WebAssembly",
+  "symbols/InteractivityPlatform/choices/WebAssembly/description": "É executado no navegador usando o WebAssembly",
+  "symbols/InteractivityPlatform/choices/Auto/displayName": "Automático (Servidor e WebAssembly)",
+  "symbols/InteractivityPlatform/choices/Auto/description": "Usa Servidor ao baixar os ativos WebAssembly e usa o WebAssembly",
```
É executado no servidor, e ao baixar os ativos WebAssembly, alterna para ele

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/auth/choices/None/description": "Sem autenticação",
+  "symbols/auth/choices/Individual/description": "Autenticação individual",
+  "symbols/auth/description": "O tipo de autenticação a ser usado",
+  "symbols/UseLocalDB/description": "Se deve usar LocalDB em vez de SQLite. Esta opção só se aplica se --auth Individual for especificado.",
```
Usar LocalDB em vez de SQLite. Esta opção só se aplica se --auth Individual for especificado.

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/UseLocalDB/description": "Se deve usar LocalDB em vez de SQLite. Esta opção só se aplica se --auth Individual for especificado.",
+  "symbols/AllInteractive/displayName": "_Habilitar renderização interativa globalmente no site",
+  "symbols/AllInteractive/description": "Configura se todas as páginas devem ser interativas aplicando um modo de renderização interativo no nível superior. Se for falso, as páginas usarão a renderização do servidor estático por padrão e poderão ser marcadas como interativas por página ou por componente.",
+  "symbols/NoHttps/description": "Se deve desligar o HTTPS. Essa opção só se aplica se Individual não for usado para --auth.",
```
Desativar HTTPS. Essa opção só se aplica se Individual não for usado para --auth.

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/AllInteractive/displayName": "_Habilitar renderização interativa globalmente no site",
+  "symbols/AllInteractive/description": "Configura se todas as páginas devem ser interativas aplicando um modo de renderização interativo no nível superior. Se for falso, as páginas usarão a renderização do servidor estático por padrão e poderão ser marcadas como interativas por página ou por componente.",
+  "symbols/NoHttps/description": "Se deve desligar o HTTPS. Essa opção só se aplica se Individual não for usado para --auth.",
+  "symbols/UseProgramMain/displayName": "Não use ins_truções de nível superior",
```
Não utilizar instruções de nível superior

HClausing commented on src/MudBlazorWebTemplate/.template.config/localize/templatestrings.pt-BR.json:
```diff
@@ -0,0 +1,44 @@
+  "symbols/AllInteractive/description": "Configura se todas as páginas devem ser interativas aplicando um modo de renderização interativo no nível superior. Se for falso, as páginas usarão a renderização do servidor estático por padrão e poderão ser marcadas como interativas por página ou por componente.",
+  "symbols/NoHttps/description": "Se deve desligar o HTTPS. Essa opção só se aplica se Individual não for usado para --auth.",
+  "symbols/UseProgramMain/displayName": "Não use ins_truções de nível superior",
+  "symbols/UseProgramMain/description": "Se deve gerar uma classe de Programa explícita e um método principal em vez de instruções de nível superior.",
```
Define se deve ser gerara uma classe de Program explícita e um método principal em vez de instruções de nível superior.

HClausing commented:
> @HClausing this is a 1:1 copy from https://github.com/dotnet/aspnetcore/tree/release/8.0/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp If you want this changed suggest it there so we can "resync" it again.

I'm not against a 1:1 copy, but I guess it's not well written on original template source.

ScarletKuro commented:
> > @HClausing this is a 1:1 copy from https://github.com/dotnet/aspnetcore/tree/release/8.0/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp If you want this changed suggest it there so we can "resync" it again.
> 
> I'm not against a 1:1 copy, but I guess it's not well written on original template source.

Is the Portuguese translation in the official template subpar? I wonder if it would make sense to make a PR with suggested changes into the aspnetcore then.

henon commented:
@HClausing did you try this template?

ScarletKuro commented:
I haven't had the opportunity to test this yet, as I'll only be able to do so after the release of .NET 8
It seems that some tweaking might be required. I see the authentication option added, but it appears to be using the native Blazor components (such as the navbar). It would be more logical to use MudBlazor components. However someone reported that SignInManager not functioning correctly with MudBlazor. There are a lot of open questions.

mahald commented:
> I haven't had the opportunity to test this yet, as I'll only be able to do so after the release of .NET 8 It seems that some tweaking might be required. I see the authentication option added, but it appears to be using the native Blazor components (such as the navbar). It would be more logical to use MudBlazor components. However someone reported that SignInManager not functioning correctly with MudBlazor. There are a lot of open questions.

The NavBar is from MudBlazor. But if you choose the option with Auth this adds bootstrap and uses it since this is the original from microsoft. it's a 1:1 copy from Microsofts "Blazor Web App" https://github.com/dotnet/aspnetcore/tree/release/8.0/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp with NavMenu / Bar, Counter, Wheater, Home using MudBlazor. it supports all options `dotnet new mudblazorwebapp -int (Auto|Server|WebAssembly)` per page and global: `dotnet new mudblazorwebapp -int (Auto|Server|WebAssembly) -ai`

mahald commented:
and you can see it in Visual Studio with the same options as the original

![image](https://github.com/MudBlazor/Templates/assets/6714894/142b85d5-90f2-4acc-98f7-b4c2a5e08f50)

HClausing commented:
> @HClausing did you try this template?

Actually not yet. I was just taking a look on open issues and PR's from MudBlazor, looking for something that I could help in the short term

mahald commented:
it's worth looking at this template: https://github.com/MudBlazor/MudBlazor/discussions/7430#discussioncomment-7386079 
my reply to it: https://github.com/MudBlazor/MudBlazor/discussions/7430#discussioncomment-7391456

henon commented:
@mahald We want to avoid that people who don't know what they are doing try this out with options that don't work with MudBlazor and later create numerous issues. This might even reflect badly on MudBlazor creating rumors that it doesn't work with NET8.0. So if the other template you mentioned is better suited for MudBlazor in your opinion (and has no noop-traps) maybe you could PR that instead? It is your decision.

mahald committed:
Update .NET 8 Web App Template

* Auth works in any Interactivity Mode
* Improve Menu for per page Interactivity
* Improve 'Per page' interactivity by providing the MudThemeProvider gloabally with UsePopoverProvider=false  (CascadingValue)
* Update Readme for long version with dotnet new command for more clarity

mahald commented:
The template has been updated and now functions flawlessly in any mode, aligning with the current release/8.0 template. Enhancements have been made to the 'Per Page' interactivity, ensuring it remains consistent with the original template. Additionally, I have included guidance on the Home page to help navigate the most significant pitfalls. With these updates, I believe we are well-prepared for the .NET 8 launch in two weeks. At that point, we will just need to update all dependencies to version 8.* (note that this is not applicable for pre-release versions).

mahald commented:
Tested:
```

dotnet new install .\ --force
c:
MD C:\temp\tests\mudTemplates
CD C:\temp\tests\mudTemplates

RD /S /Q interactivityNone

RD /S /Q interactivityAuto
RD /S /Q interactivityServer
RD /S /Q interactivityWebAssembly

RD /S /Q interactivityAuto_allInteractive
RD /S /Q interactivityServer_allInteractive
RD /S /Q interactivityWebAssembly_allInteractive

RD /S /Q interactivityAuto_auth
RD /S /Q interactivityAuto_allInteractive_auth

RD /S /Q interactivityAuto_useMain

dotnet new mudblazorwebapp --interactivity None --output C:\temp\tests\mudTemplates\interactivityNone

dotnet new mudblazorwebapp --interactivity Auto --output C:\temp\tests\mudTemplates\interactivityAuto
dotnet new mudblazorwebapp --interactivity Server --output C:\temp\tests\mudTemplates\interactivityServer
dotnet new mudblazorwebapp --interactivity WebAssembly --output C:\temp\tests\mudTemplates\interactivityWebAssembly

dotnet new mudblazorwebapp --interactivity Auto --output C:\temp\tests\mudTemplates\interactivityAuto_allInteractive --all-interactive
dotnet new mudblazorwebapp --interactivity Server --output C:\temp\tests\mudTemplates\interactivityServer_allInteractive --all-interactive
dotnet new mudblazorwebapp --interactivity WebAssembly --output C:\temp\tests\mudTemplates\interactivityWebAssembly_allInteractive --all-interactive

dotnet new mudblazorwebapp --interactivity Auto --output C:\temp\tests\mudTemplates\interactivityAuto_auth --auth Individual
dotnet new mudblazorwebapp --interactivity Auto --output C:\temp\tests\mudTemplates\interactivityAuto_allInteractive_auth --all-interactive --auth Individual

dotnet new mudblazorwebapp --interactivity Auto --output C:\temp\tests\mudTemplates\interactivityAuto_useMain --use-program-main

cd %~dp0
```

henon commented:
Wow, you did a great job of explaining the different options in the template. Love it!

Garderoben commented:
@mahald if i would sit down and convert all the none mudblazor stuff to mudblazor is there any pitfalls here? Any pages that still does not run "Blazor"? i remember me and @just-the-benno had a fix for this when working on the podcast app.

mahald commented:
> @mahald if i would sit down and convert all the none mudblazor stuff to mudblazor is there any pitfalls here? Any pages that still does not run "Blazor"? i remember me and @just-the-benno had a fix for this when working on the podcast app.


Microsoft handles authentication in SSR mode with Form and EditForm. I believe MudBlazor would require interactivity for that. I've experienced issues with constant redirection when attempting to implement interactivity in Auth Pages. Thus, it might require additional effort.

For now, in my project, I’ve synchronized the colors using some CSS:

```
:root {
    --bs-primary: var(--mud-palette-primary) !important;
}

.btn-primary {
    --bs-btn-bg: var(--mud-palette-primary) !important;
    --bs-btn-hover-bg: var(--mud-palette-secondary) !important;
}
```

I was considering omitting all the authentication components, introducing .AddOpenApi, and incorporating SwaggerUI in DevMode. This way, developers can simply test the API using SwaggerUI and proceed to implement the necessary functions."

mahald commented:
Utilizing the Auth components also reveals the potential for blending different frameworks seamlessly. In my test project, I applied `text-transform: uppercase;` to the CSS, making the Bootstrap elements in the Auth component virtually indistinguishable from the MudBlazor elements; their styles are strikingly similar, and the buttons appear identical to me.

I'm considering creating an app.css file for now, with the following content:

```css
:root {
    --bs-primary: var(--mud-palette-primary) !important;
}

.btn-primary {
    text-transform: uppercase;
    --bs-btn-bg: var(--mud-palette-primary) !important;
    --bs-btn-hover-bg: var(--mud-palette-secondary) !important;
}
```

mahald committed:
.NET 8 WebApp - match color for auth components

css to take mudBlazor colors for bootstrap, also had to update bootstrap to make this work.

starkguerraj commented:
Was testing this template out and get the following error.

Appears that removing 

<MudPopoverProvider></MudPopoverProvider> from MyMudProviders.razor resolves the issue.

Template was generated using auto interactivity mode, including samples with the global scope.

```
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Duplicate MudPopoverProvider detected. Please ensure there is only one provider, or disable this warning with PopoverOptions.ThrowOnDuplicateProvider.
System.InvalidOperationException: Duplicate MudPopoverProvider detected. Please ensure there is only one provider, or disable this warning with PopoverOptions.ThrowOnDuplicateProvider.
   at MudBlazor.MudPopoverProvider.OnAfterRenderAsync(Boolean firstRender)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
```

SigurdEEspersen commented:
If this template is how we're getting MudBlazor compatible with the new features of .NET 8.0. When can we expect this PR to be merged?

AmBplus commented:
When We Seen The Offical Template Of MudBlazor For NET8.0

henon commented:
Thank you @mahald ! @Garderoben said he'll publish it as soon as he finds time.
